src/json.hpp

```diff
@@ -2702,32 +2644,51 @@ class lexer
     // input management
     /////////////////////
 
-    /// reset yytext
+    /// reset yytext; current character is beginning of token
     void reset() noexcept
     {
-        yylen = 0;
-        start_pos = chars_read - 1;
+        yytext.clear();
+        token_string.clear();
+        token_string.push_back(static_cast<char>(current));
     }
 
-    /// get a character from the input
+    /*
+    @brief get next character from the input
+
+    This function provides the interface to the used input adapter. It does
+    not throw in case the input reached EOF, but returns a -'ve valued
+    `std::char_traits<char>::eof()` in that case.  Stores the scanned characters
+    for use in error messages.
+
+    @return character read from the input
+    */
     int get()
     {
         ++chars_read;
-        return next_unget ? (next_unget = false, current)
-               : (current = ia->get_character());
+        current = ia->get_character();
+        if (JSON_LIKELY(current >= 0)) // faster than: != std::char_traits<char>::eof()))
```

Comment:
Probably not safe.  See: https://stackoverflow.com/questions/24482728/return-value-of-istreamget

In fact, all of the assumed `int get()` and `int get_character()` is also not strictly correct.  It should be istream::int_type.  And, we should be using istream::traits_type::eq_int_type to detect equality w/ eof(), and istream::traits_type::to_char_type to convert to a char (not ... & 0xFF).

So, I agree -- perhaps we should clean up the char handling a bit to bring it more into strict compliance.  This may cost us the 1-2% performance increase we saw by just testing for negative values instead of == eof() specifically.